libcalico-go/lib/backend/model/keys.go

```diff
@@ -193,18 +189,257 @@ func IsListOptionsLastSegmentPrefix(listOptions ListInterface) bool {
+			return nil
+		}
+		switch parts[2] {
+		case "v2":
+			switch parts[3] {
+			case "assignment":
+				return BlockListOptions{}.KeyFromDefaultPath(path)
+			case "handle":
+				if len(parts) > 5 {
+					return nil
+				}
+				return IPAMHandleKey{
+					HandleID: parts[4],
+				}
+			case "host":
+				return BlockAffinityListOptions{}.KeyFromDefaultPath(path)
+			}
+		}
+	case "resources":
+		switch parts[2] {
+		case "v3":
+			// v3 resource keys strictly require the leading slash.
+			if len(parts) < 6 || parts[3] != "projectcalico.org" || path[0] != '/' {
+				return nil
+			}
+			switch len(parts) {
+			case 6:
+				ri, ok := resourceInfoByPlural[unescapeName(parts[4])]
+				if !ok {
+					log.Warnf("(BUG) unknown resource type: %v", path)
+					return nil
+				}
+				if namespace.IsNamespaced(ri.kind) {
+					log.Warnf("(BUG) Path is a global resource, but resource is namespaced: %v", path)
+					return nil
+				}
+				log.Debugf("Path is a global resource: %v", path)
+				return ResourceKey{
+					Kind: ri.kind,
+					Name: unescapeName(parts[5]),
+				}
+			case 7:
+				ri, ok := resourceInfoByPlural[unescapeName(parts[4])]
+				if !ok {
+					log.Warnf("(BUG) unknown resource type: %v", path)
+					return nil
+				}
+				if !namespace.IsNamespaced(ri.kind) {
+					log.Warnf("(BUG) Path is a namespaced resource, but resource is global: %v", path)
+					return nil
+				}
+				log.Debugf("Path is a namespaced resource: %v", path)
+				return ResourceKey{
+					Kind:      ri.kind,
+					Namespace: unescapeName(parts[5]),
+					Name:      unescapeName(parts[6]),
+				}
+			}
+		}
+	case "felix":
+		if len(parts) < 4 {
+			return nil
+		}
+		switch parts[2] {
+		case "v1":
+			switch parts[3] {
+			case "host":
+				if len(parts) != 7 || parts[5] != "endpoint" {
+					return nil
+				}
+				return HostEndpointStatusKey{
+					Hostname:   parts[4],
+					EndpointID: unescapeName(parts[6]),
+				}
+			}
+		case "v2":
+			if len(parts) < 7 {
+				return nil
+			}
+			if parts[4] != "host" {
+				return nil
+			}
+			switch parts[6] {
+			case "status":
+				return ActiveStatusReportListOptions{}.KeyFromDefaultPath(path)
+			case "last_reported_status":
+				return LastStatusReportListOptions{}.KeyFromDefaultPath(path)
+			case "workload":
+				return WorkloadEndpointStatusListOptions{}.KeyFromDefaultPath(path)
+			}
+		}
+	}
+	log.Debugf("Path is unknown: %v", path)
+	return nil
+}
+
+// OldKeyFromDefaultPath is the old, (slower) implementation of KeyFromDefaultPath.  It is kept to allow
+// fuzzing the new version against it.  Parses the default path representation of a key into one
+// of our <Type>Key structs.  Returns nil if the string doesn't match one of
+// our key types.
+func OldKeyFromDefaultPath(path string) Key {
 	if m := matchWorkloadEndpoint.FindStringSubmatch(path); m != nil {
 		log.Debugf("Path is a workload endpoint: %v", path)
 		return WorkloadEndpointKey{
-			Hostname:       m[1],
+			Hostname:       unescapeName(m[1]),
```

Comment:
I think the old code was bugged here: we apply encoding when converting a key to a path but not the other way around.  In practice, validation prevents encoding from being needed but, given a bad key added to etcd we'd be unable to round trip it with the old code.

Reply:
I changed it so the Fuzzer would stop finding this problem; couldn't think of a good way to avoid that issue in the fuzz test.

Reply:
OK, agree that that makes sense.